projects/pgai/pgai/vectorizer/parsing.py

```diff
@@ -24,32 +36,91 @@ class ParsingAuto(BaseModel):
+                return ParsingPyMuPDF(implementation="pymupdf").parse(row, payload)
+
+            return ParsingDocling(implementation="docling").parse(row, payload)
         else:
             return payload
 
 
-class ParsingPyMuPDF(BaseModel):
-    implementation: Literal["pymupdf"]
+class BaseDocumentParsing(BaseModel, ABC):
+    """Base class for document parsing implementations."""
+
+    implementation: str
+
+    def parse(self, row: dict[str, Any], payload: LoadedDocument | str) -> str:
+        """
+        Parse a document payload into a string representation.
+
+        Args:
+            row: Metadata about the document. The whole actual db row
+            payload: Either a LoadedDocument or raw string content
 
-    def parse(self, row: dict[str, Any], payload: LoadedDocument | str) -> str:  # noqa: ARG002
+        Returns:
+            Parsed string content. Markdown preferable.
+
+        Raises:
+            ValueError: If payload type is invalid or file type cannot be determined
+        """
         if isinstance(payload, str):
             raise ValueError(
-                "Column content must be a document to be parsed by pymupdf "
-                "use parsing_auto or parsing_none instead"
+                f"Column content must be a document to be parsed by "
+                f"{self.implementation}, use parsing_auto or parsing_none instead"
             )
 
         if payload.file_type is None:
             raise ValueError("No file extension could be determined")
 
-        if payload.file_type in ["txt", "md"]:  # type: ignore
-            # no parsing is needed
+        if payload.file_type in ["txt", "md"]:
             return payload.content.getvalue().decode("utf-8")
 
+        return self.parse_doc(row, payload)
+
+    @abstractmethod
+    def parse_doc(self, row: dict[str, Any], payload: LoadedDocument) -> str:
+        """
+        Parse a binary document into a string representation, Markdown preferable.
+        Must be implemented by subclasses.
+        """
+
+
+class ParsingPyMuPDF(BaseDocumentParsing):
+    """Document parsing implementation using PyMuPDF."""
+
+    implementation: Literal["pymupdf"]  # type: ignore[reportIncompatibleVariableOverride]
+
+    @override
+    def parse_doc(self, row: dict[str, Any], payload: LoadedDocument) -> str:  # noqa: ARG002
         with pymupdf.open(
             stream=payload.content, filetype=payload.file_type
         ) as pdf_document:  # type: ignore
-            # Convert to markdown using pymupdf4llm
-            md_text = pymupdf4llm.to_markdown(pdf_document)  # type: ignore
+            return pymupdf4llm.to_markdown(pdf_document)  # type: ignore
+
+
+class ParsingDocling(BaseDocumentParsing):
+    """Document parsing implementation using Docling."""
+
+    implementation: Literal["docling"]  # type: ignore[reportIncompatibleVariableOverride]
+    cache_dir: Path | str = Path.home().joinpath(".cache/docling/models")
```

Comment:
In the future we could make this configurable via env var. Just using the default value from docling atm.